cerberus/tests/test_validation.py

```diff
@@ -1838,3 +1838,71 @@ def test_contains(constraint):
         errors.MISSING_MEMBERS
     ].info[0]
     assert any(x in missing_actors for x in ('Eric Idle', 'Terry Gilliam'))
+
+
+def test_require_all_simple():
+    schema = {'foo': {'type': 'string'}}
+    validator = Validator(require_all=True)
+    assert_fail(
+        {},
+        schema,
+        validator,
+        error=('foo', ('foo', 'required'), errors.REQUIRED_FIELD, True),
```

Comment:
as a start replace the second tuple value here with `'__require_all__'`. as far as i can oversee the situation, the `schema_path` variable in the `Validator._error` method must then be figured out with regards to the presence of the `required` rule for the submitted `field` in the schema when an `errors.REQUIRED_FIELD` is processed.

Reply:
That is obvious... and this is test, I mean, at first we need to fix code, so it will result `__require_all__` in error schema path and only after tests should be fixed. The problem is to fix `validator.py` and replace `required` with `__require_all__` in `__validate_requried_fields` in some cases.

Reply:
> and this is test, I mean, at first we need to fix code

following the paradigm of test-driven development, one first formulates assertions in a test.

`__validate_required_fields` calls `_error` which only takes a reference to the field and an error definition as argument, hence the detemination of the correct (pseudo-) schema path has to happen in the latter method.

Reply:
> following the paradigm of test-driven development, one first formulates assertions in a test.

Ok, but this does not help me at all. TDD is for developers, this PR won't be accepted without tests and implementation, so there is no reason to order changes.
> hence the detemination of the correct (pseudo-) schema path has to happen in the latter method.

I don't get it. We have `__validate_unknown_fields`, which calls `_error` with some preparation and we have `__validate_required_fields`, which calls `_error` without something. Why this cases are not equal? What is "later method" in this case?

Reply:
i don't get it.

> TDD is for developers, this PR won't be accepted without tests and implementation, so there is no reason to order changes.

afaict you are developing a feature here, hence you're in the role of a developer. i don't care how you work if the result is fine from a variety of angles, one of these are tests that define expected behaviour and are passing. at the moment the test defines an undesired behaviour.

> Why this cases are not equal?

`__validate_unknown_fields` uses a 'child validator' that handles the schema path (that may eventually be used in `_error`).

> `__validate_required_fields`, which calls `_error` without something

no. that never happens.

> What is "later method" in this case?

the latter that was mentioned in the previous part in the sentence, `_error`.

Reply:
> if the result is fine from a variety of angles, one of these are tests that define expected behaviour and are passing. at the moment the test defines an undesired behaviour.

That is the point. I understand what behaviour we need, but I don't understand how to do it. And advice to "start with test" sound like a joke. :smile: 

> the latter that was mentioned in the previous part in the sentence, `_error`.

You suggest to add `__require_all__` to `schema_path` in `_error` method, am I right?

Reply:
> And advice to "start with test" sound like a joke.

neither did i intend to be funny with that statement nor can i find traces of humour in it.

so far, my successes with that approach were really fun, but not laughable.

> You suggest to add `__require_all__` to schema_path in `_error` method, am I right?

more specificly: `__require_all__` *is* the `schema_path` when the error was defined by the `require_all` option of the root validator.